data/tutorials/language/0lg_10_polymorphic_variants.md

```diff
@@ -0,0 +1,713 @@
+```ocaml
+# type a = A;;
+type a = A
+# type b = A;;
+type b = A
+# function A -> true;;
+- : b -> bool = <fun>
+# function (A : a) -> true;;
+- : a -> bool = <fun>
+# (function A -> true : a -> bool);;
+- : a -> bool = <fun>
+```
+
+Without this, previously entered ones are no longer reachable. This can be worked around using modules. This is explained in [Reusing Field Names](https://dev.realworldocaml.org/records.html#scrollNav-3)
+) of the _Real World OCaml_ book written by Yaron Minsky and Anil Madhavapeddy.
+
+This problem never happens with polymorphic variants. When a tag appears several times in an expression, it must be with the same type, that's the only restriction. This makes polymorphic variants very handy when dealing with multiple sum types sharing constructors.
+
+### Data Type Sharing Between Modules
+
+Using the same type in two different modules can be done in several ways:
+- Having a dependency, either direct or shared
+- Turn the dependent module into a functor and inject the dependence as a parameter
+
+The polymorphic variants provide an additional alternative. This was proposed by Jacques Garrigue in his seminal paper “Programming with Polymorphic Variants” (ACM SIGPLAN Workshop on ML, October 1998):
+> You [...] define the same [polymorphic variant] type in both [modules], and since these are only type abbreviations, the two definitions are compatible. The type system checks the structural equality when you pass a value from one [module] to the other.
+
+### Error Handling Using the `result` Type
+
+The [Error Handling](/docs/error-handling) guide details possible ways to handle errors in OCaml. Among various mechanisms, the `result` type, when used as a monad, provides a powerful means to handle errors. Refer to the guide and documentation on this type to learn how to use it. In this section, we discuss why using polymorphic variants to carry error values can be beneficial.
+
+Let's consider this exception-raising code:
+```ocaml
+# let f_exn m n =
+    let open List in
+    let u = init m Fun.id |> map (fun n -> n * n) in
+    nth u n;;
+val f_exn : int -> int -> int = <fun>
+```
+
+The following is an attempt at translating `f_exn` with `result` values instead of exceptions. It is using the `Result.bind` instead of `|>`:
+```ocaml
+# type init_error = Negative_length;;
+
+# let init n f = try Ok (List.init n f) with
+    | Invalid_argument _ -> Error Negative_length;;
+val init : int -> (int -> 'a) -> ('a list, init_error) result = <fun>
+
+# type nth_error = Too_short of int * int | Negative_index of int;;
+type nth_error = Too_short of int * int | Negative_index of int
+
+# let nth u i = try Ok (List.nth u i) with
+    | Invalid_argument _ -> Error (Negative_index i)
+    | Failure _ -> Error (Too_short (List.length u, i))
+val nth : 'a list -> int -> ('a, nth_error) result = <fun>
+
+# let f_res m n =
+    let* u = init m Fun.id in
```

Comment:
I think `let*` is not defined in this snippet, and when defining it in the snippet below it may be worth linking to some text about the `let*` syntax(?)

Reply:
You are right, the definition of let should move up. Links are also needed you are right